projects/sampling/executor.py

```diff
@@ -0,0 +1,185 @@
+    output_artifact = artifact_utils.get_single_instance(output_dict[component.UNDERSAMPLER_OUTPUT_KEY])
+
+    if copy_others:
+      output_artifact.split_names = input_artifact.split_names
+    else:
+      output_artifact.split_names = artifact_utils.encode_split_names(splits)
+
+    # Fetch the input uri for each split
+    split_data = {}
+    for split in artifact_utils.decode_split_names(input_artifact.split_names):
+      uri = artifact_utils.get_split_uri([input_artifact], split)
+      split_data[split] = uri
+
+    for split, uri in split_data.items():
+      if split in splits:  # Undersampling split
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        os.mkdir(output_dir)
+        self.undersample(uri, label, shards, keep_classes, os.path.join(output_dir, f"Split-{split}"),)
+      elif copy_others:  # Copy the other split if copy_others is True
+        input_dir = uri
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        os.mkdir(output_dir)
+        for filename in fileio.listdir(input_dir):
+          input_uri = os.path.join(input_dir, filename)
+          output_uri = os.path.join(output_dir, filename)
+          io_utils.copy_file(src=input_uri, dst=output_uri, overwrite=True)
+
+  def undersample(self, uri, label, shards, keep_classes, output_dir):
+    """Function that actually undersamples the given split.
+
+    Args:
+      uri: The input uri for the specific split of the input example artifact.
+      label: The name of the column containing class names to undersample by.
+      shards: The number of files that each undersampled split should
+      contain. Default 0 is Beam's tfrecordio function's default.
+      keep_classes: A list determining which classes that we should
+      not undersample. Defaults to None.
+      output_dir: The output directory for the split of the output artifact.
+
+    Returns:
+      None
+    """
+
+    def generate_elements(x):
+      class_label = None
+      parsed = tf.train.Example.FromString(x.numpy())
+      if parsed.features.feature[label].int64_list.value:
+        val = parsed.features.feature[label].int64_list.value
+        if len(val) > 0:
+          class_label = val[0]
+      else:
+        val = parsed.features.feature[label].bytes_list.value
+        if len(val) > 0:
+          class_label = val[0].decode()
+      return (class_label, parsed)
+
+    def sample(key, value, side=0):
+      for item in random.sample(value, side):
+        yield item
+
+    def filter_null(item, keep_null=False, null_vals=None, pr=False):
+      if pr:
+        print(item)
+      keep = (not item[0]) or item[0] == 0  # Note that 0 is included in this filter!
+      if null_vals:
+        keep = keep or str(item[0]) in null_vals
+      keep ^= not keep_null  # null is True if we want to keep nulls
+      if keep:
+        return item[1]
+  
+    files = [os.path.join(uri, name) for name in os.listdir(uri)]
```

Comment:
Again `os.listdir()` won't work if you're using GCS as a storage bucket for your metadata. I fixed this by changing 
``` 
files = [os.path.join(uri, name) for name in os.listdir(uri)]
dataset = tf.data.TFRecordDataset(files, compression_type="GZIP")
```
into just
```
dataset = tf.data.TFRecordDataset(files, compression_type="GZIP")
```